### PR TITLE
First pass at streaming bodies from fetch()

### PR DESCRIPTION
**Pull request by ry**

### CODE REVIEW

File: js/fetch.ts [hunk 1]
```diff
@@ -97,59 +99,127 @@ export const Headers = DomIterableMixin<string, string, typeof HeadersBase>(
...
+    });
+  }
+
+  async formData(): Promise<domTypes.FormData> {
+    notImplemented();
+    return {} as domTypes.FormData;
+  }
+
+  // tslint:disable-next-line:no-any
+  async json(): Promise<any> {
```
Comment (piscisaureus) [inline js/fetch.ts]: Use unknown.

File: js/fetch.ts [hunk 2]
```diff
@@ -97,59 +99,127 @@ export const Headers = DomIterableMixin<string, string, typeof HeadersBase>(
...
+  constructor(readonly rid: number, readonly contentType: string) {
+    this._bodyWaiter = createResolvable();
+  }
+
+  private async _bufferBody(): Promise<void> {
+    if (!this._buffering) {
+      this._buffering = true;
+      try {
+        while (true) {
+          const b = new Uint8Array(64 * 1024);
```
Comment (piscisaureus) [inline js/fetch.ts]: Is there no content-length header so we can just preallocate a buffer with the right size?

File: js/fetch.ts [hunk 3]
```diff
@@ -97,59 +99,127 @@ export const Headers = DomIterableMixin<string, string, typeof HeadersBase>(
...
+          const r = await read(this.rid, b);
+          if (r.nread > 0) {
+            this._chunks.push(b.subarray(0, r.nread));
+          }
+          if (r.eof) {
+            break;
+          }
+        }
+        this._data = concatUint8Array(this._chunks);
```
Comment (piscisaureus) [inline js/fetch.ts]: It'd be more efficient to skip this step if there is only 1 chunk.

File: js/fetch.ts [hunk 4]
```diff
@@ -97,59 +99,127 @@ export const Headers = DomIterableMixin<string, string, typeof HeadersBase>(
...
+        this._chunks = [];
+        this._bodyWaiter.resolve(this._data!.buffer as ArrayBuffer);
```
Comment (piscisaureus) [inline js/fetch.ts]: I don't think you need ! here, nor that type assertion.

File: js/fetch.ts [hunk 5]
```diff
@@ -97,59 +99,127 @@ export const Headers = DomIterableMixin<string, string, typeof HeadersBase>(
...
-    notImplemented();
-    return {} as domTypes.FormData;
+    return this.body.formData();
   }
 
-  async json(): Promise<object> {
-    const text = await this.text();
-    return JSON.parse(text);
+  // tslint:disable-next-line:no-any
+  async json(): Promise<any> {
```
Comment (piscisaureus) [inline js/fetch.ts]: the type to use here would be `unknown`.

File: src/resources.rs [hunk 6]
```diff
@@ -189,6 +194,16 @@ pub fn add_tcp_stream(stream: tokio::net::TcpStream) -> Resource {
   Resource { rid }
 }
 
+pub fn add_hyper_body(body: hyper::Body) -> Resource {
+  let rid = new_rid();
+  let mut tg = RESOURCE_TABLE.lock().unwrap();
+
+  let body = HttpBody::from(body);
+  let r = tg.insert(rid, Repr::HttpBody(body));
```
Comment (piscisaureus) [inline src/resources.rs]: Cool. Where does the body get removed from the table?

File: js/fetch.ts (hunk 1 above)
Comment (ry) [reply js/fetch.ts]: That results in this error:
```
/Users/rld/src/deno/js/fetch_test.ts:8:15 - error TS2571: Object is of type 'unknown'.

8   assertEqual(json.name, "deno");
                ~~~~
```

File: js/fetch.ts (hunk 4 above)
Comment (ry) [reply js/fetch.ts]: Removed the ! - good catch

If I remove the cast, I get:
```
[!] (rpt2 plugin) Error: ../../js/fetch.ts:131:34 - error TS2345: Argument of type 'ArrayBuffer | SharedArrayBuffer' is not assignable to parameter of type 'ArrayBuffer | PromiseLike<ArrayBuffer> | undefined'.
  Type 'SharedArrayBuffer' is not assignable to type 'ArrayBuffer | PromiseLike<ArrayBuffer> | undefined'.
    Type 'SharedArrayBuffer' is not assignable to type 'PromiseLike<ArrayBuffer>'.
      Property 'then' is missing in type 'SharedArrayBuffer'.

131         this._bodyWaiter.resolve(this._data.buffer);
```

File: js/fetch.ts (hunk 2 above)
Comment (ry) [reply js/fetch.ts]: This whole bit is preliminary and will be replaced with a general purpose method of buffering any readable object. 
See https://golang.org/pkg/bytes/#Buffer.ReadFrom and WIP https://github.com/ry/deno/blob/buffer/js/buffer.ts
I've added a TODO comment that discusses this.

File: js/fetch.ts (hunk 3 above)
Comment (ry) [reply js/fetch.ts]: Preliminary, added TODO about buffer code.

File: js/fetch.ts (hunk 5 above)
Comment (ry) [reply js/fetch.ts]: Same error: https://github.com/denoland/deno/pull/1102#discussion_r228714933

File: src/resources.rs (hunk 6 above)
Comment (ry) [reply src/resources.rs]: Uh - that's a very good point and potential glaring oversight on my part... I need to investigate.

File: js/fetch.ts (hunk 1 above)
Comment (kitsonk) [reply js/fetch.ts]: `unknown` is assignable to any type... it just keeps people from using it without putting some sort of type cast on it.  You could simple do:

```
const json: any = await fetch(...).json();
// or
const json: { [v: string]: string; } = await fetch(...).json();
```

But considering we haven't forced strict mode on Deno user land, this seems like an overly strict construct to for on users now.

**Comment by ry** [comment]
@piscisaureus PTAL

File: src/resources.rs (hunk 6 above)
Comment (ry) [reply src/resources.rs]: Fixed - added a close method.

File: src/http_body.rs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+  fn read(&mut self, _buf: &mut [u8]) -> io::Result<usize> {
+    unimplemented!();
+  }
+}
+
+impl AsyncRead for HttpBody {
+  fn poll_read(&mut self, buf: &mut [u8]) -> Poll<usize, io::Error> {
+    match self.chunk.take() {
+      Some(chunk) => {
+        debug!(
```
Comment (piscisaureus) [inline src/http_body.rs]: How do you get this debug to print out?

File: js/fetch.ts [hunk 8]
```diff
@@ -1,73 +1,146 @@
...
+  async blob(): Promise<domTypes.Blob> {
+    const arrayBuffer = await this.arrayBuffer();
+    return new DenoBlob([arrayBuffer], {
+      type: this.contentType
+    });
+  }
+
+  async formData(): Promise<domTypes.FormData> {
+    notImplemented();
+    return {} as domTypes.FormData;
```
[the remark this answers is not in the document]
Comment (ry) [reply js/fetch.ts]: Fixed.

**Pull request merged**